File: langs/dotnet/src/Dotlyte/Coercion.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Dotlyte;

/// <summary>
/// Type coercion engine.
/// </summary>
public static partial class Coercion
{
    private static readonly HashSet<string> NullValues = new(StringComparer.OrdinalIgnoreCase)
        { "null", "none", "nil", "" };

    private static readonly HashSet<string> TrueValues = new(StringComparer.OrdinalIgnoreCase)
        { "true", "yes", "1", "on" };

    private static readonly HashSet<string> FalseValues = new(StringComparer.OrdinalIgnoreCase)
        { "false", "no", "0", "off" };

    /// <summary>
    /// Coerce a string value to its proper .NET type.
    /// </summary>
    public static object? Coerce(object? value)
    {
        if (value is not string str)
            return value;

        var trimmed = str.Trim();
        var lower = trimmed.ToLowerInvariant();

        // Null
        if (NullValues.Contains(lower))
            return null;

        // Boolean true
        if (TrueValues.Contains(lower))
            return true;

        // Boolean false
        if (FalseValues.Contains(lower))
            return false;

        // Integer
        if (Regex.IsMatch(trimmed, @"^-?\d+$"))
        {
            if (long.TryParse(trimmed, NumberStyles.Integer, CultureInfo.InvariantCulture, out var intVal))
                return intVal;
        }

        // Float
        if (trimmed.Contains('.') && Regex.IsMatch(trimmed, @"^-?\d+\.\d+$"))
        {
            if (double.TryParse(trimmed, NumberStyles.Float, CultureInfo.InvariantCulture, out var floatVal))
                return floatVal;
        }

        // List (comma-separated)
        if (trimmed.Contains(','))
        {
            return trimmed.Split(',')
                .Select(item => Coerce(item.Trim()))
                .ToList();
        }

        return trimmed;
    }

    /// <summary>
    /// Recursively coerce all string values in a dictionary.
    /// </summary>
    public static Dictionary<string, object?> CoerceDictionary(Dictionary<string, object?> data)
    {
        var result = new Dictionary<string, object?>();

        foreach (var (key, value) in data)
        {
            result[key] = value switch
            {
                Dictionary<string, object?> dict => CoerceDictionary(dict),
                string s => Coerce(s),
                _ => value,
            };
        }

        return result;
    }
}

File: langs/php/src/Coercion.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Type coercion engine.
 */
final class Coercion
{
    private const NULL_VALUES = ['null', 'none', 'nil', ''];
    private const TRUE_VALUES = ['true', 'yes', '1', 'on'];
    private const FALSE_VALUES = ['false', 'no', '0', 'off'];

    /**
     * Coerce a string value to its proper PHP type.
     */
    public static function coerce(mixed $value): mixed
    {
        if (!is_string($value)) {
            return $value;
        }

        $trimmed = trim($value);
        $lower = strtolower($trimmed);

        // Null
        if (in_array($lower, self::NULL_VALUES, true)) {
            return null;
        }

        // Boolean true
        if (in_array($lower, self::TRUE_VALUES, true)) {
            return true;
        }

        // Boolean false
        if (in_array($lower, self::FALSE_VALUES, true)) {
            return false;
        }

        // Integer
        if (preg_match('/^-?\d+$/', $trimmed)) {
            return (int) $trimmed;
        }

        // Float
        if (str_contains($trimmed, '.') && preg_match('/^-?\d+\.\d+$/', $trimmed)) {
            return (float) $trimmed;
        }

        // List (comma-separated)
        if (str_contains($trimmed, ',')) {
            return array_map(
                fn (string $item) => self::coerce(trim($item)),
                explode(',', $trimmed)
            );
        }

        return $trimmed;
    }

    /**
     * Recursively coerce all string values in an array.
     *
     * @param array<string, mixed> $data
     * @return array<string, mixed>
     */
    public static function coerceArray(array $data): array
    {
        $result = [];
        foreach ($data as $key => $value) {
            if (is_array($value)) {
                $result[$key] = self::coerceArray($value);
            } elseif (is_string($value)) {
                $result[$key] = self::coerce($value);
            } else {
                $result[$key] = $value;
            }
        }
        return $result;
    }
}
